### PR TITLE
Hierarchical Gravity and timestep splitting

### PR DESCRIPTION
**Pull request by sbird**
Use two timesteps in the code: one for hydro and one for gravity. Allow hydro to subsample gravity. This is what Gadget-4 does and it allows us to sometimes avoid building the full gravity tree.

Question: does BH repositioning and dynamic friction go on the hydro timescale or the gravity timescale? 
If it can be on the longer gravity timescale this would help us avoid more full treebuilds.

### CODE REVIEW

**Comment by sbird** [comment]
At the moment (on the hydro example, which is too small for to really help) this avoids moment calculation for about 50% of timesteps. Almost all timesteps where we avoid the gravity treebuild still build a tree in blackhole.

**Comment by sbird** [comment]
Rebased and fixed a bug which stopped treegrav running on PM steps. Avoid building a tree in BH if no BHs active.

**Comment by sbird** [comment]
I have made the BH feedback treewalk asymmetric. I think this is fine because flagging the particles to be swallowed is also asymmetric, but someone else should confirm this.

**Comment by sbird** [comment]
For the hydro example this now avoids building a tree (except for the gas tree) on the shortest timebin. The level up usually builds a gravity tree so it is hard to tell the real cost of the blackhole/wind tree. The cost of moment computation is now much smaller. The cost of the nodes is about the same, because we often build a gas tree and a full tree in the same timestep. This could be avoided if it makes sense on bigger problems.

**Comment by sbird** [comment]
There are a few more stars forming in the hydro example with threads than before but there will be some variance because of the slightly different Hsml values that come out of the density iteration.

File: libgadget/blackhole.c [hunk 1]
```diff
@@ -540,6 +541,21 @@ blackhole(const ActiveParticles * act, double atime, Cosmology * CP, ForceTree *
...
+     * accretion uses: all types but ONLY for repositioning potential minimum. Otherwise gas + black holes. gas + stars + BH is probably fine.
+     * gas + BH probably not enough if gas is sparse in halo.
+     * feedback uses: gas + black holes.
+     * The DM in dynamic friction and accretion doesn't really do anything, so could perhaps be removed from the treebuild later.
+     * However, we would still need a tree with gas + DM in the wind code.
+     */
+    if(!tree->tree_allocated_flag)
+    {
+        message(0, "Building tree in blackhole\n");
+        force_tree_rebuild_mask(tree, ddecomp, GASMASK + DMMASK + STARMASK + BHMASK, 0, NULL);
```
Comment (rainwoodman) [inline libgadget/blackhole.c:555]: "|"?

File: libgadget/gravshort-tree.c [hunk 2]
```diff
@@ -118,14 +117,18 @@ grav_short_tree(const ActiveParticles * act, PetaPM * pm, ForceTree * tree, doub
...
         endrun(2, "Gravtree called before tree moments computed!\n");
 
     tw->ev_label = "GRAVTREE";
     tw->visit = (TreeWalkVisitFunction) force_treeev_shortrange;
-    /* gravity applies to all particles. Including Tracer particles to enhance numerical stability. */
-    tw->haswork = NULL;
+    /* gravity applies to all gravitationally active particles.
+     * If we are a PM step then all particles are active. */
+    tw->haswork = gravshort_haswork;
+    if(!act->ActiveParticle)
```
Comment (rainwoodman) [inline libgadget/gravshort-tree.c:130]: This may be easier to read:

if(act->ActiveParticle) {
  tw->haswork = NULL;
} else {
  tw->haswork = gravshort_haswork;
}

Also, can we / shall we move the conditional into gravshort_haswork? (e.g. if !act_activeParticle return True) -- Because haswork function is supposed to be "the" conditional.

File: libgadget/run.c [hunk 3]
```diff
@@ -386,14 +387,21 @@ run(int RestartSnapNum)
...
 
-        if(All.TreeGravOn) {
+        /* Build the force tree for tree gravity.*/
+        ForceTree Tree = {0};
+        int anygravactive = MPIU_Any(Act.NumActiveGravity, MPI_COMM_WORLD);
+
+        if(is_PM || (All.TreeGravOn && anygravactive))
+            force_tree_rebuild(&Tree, ddecomp, HybridNuGrav, !pairwisestep && All.TreeGravOn, All.OutputDir);
+
+        if(All.TreeGravOn && (is_PM || anygravactive)) {
```
Comment (rainwoodman) [inline libgadget/run.c:397]: This looks puzzling -- how is this condition related to the one in line 394? Could you add a comment on when the first is True but the second is False? And looks like we will crash if the first is False but the second is True?

File: libgadget/blackhole.c (hunk 1 above)
Comment (sbird) [reply libgadget/blackhole.c:555]: Sure, I suppose. I was following the style we use for the .mask in ngbiter.

File: libgadget/gravshort-tree.c (hunk 2 above)
Comment (sbird) [reply libgadget/gravshort-tree.c:130]: Yes you are right! I will adjust the condition. We maybe should but cannot move it into gravshort_haswork because we make memory allocation decisions in treewalk_build_queue depending on whether ->haswork is NULL.

File: libgadget/blackhole.c [hunk 4]
```diff
@@ -524,8 +524,9 @@ blackhole(const ActiveParticles * act, double atime, Cosmology * CP, ForceTree *
     priv->hubble = hubble_function(CP, atime);
     priv->GravInternal = CP->GravInternal;
 
-    /* Build the queue once, since it is really 'all black holes' and similar for all treewalks*/
-    treewalk_build_queue(tw_dynfric, act->ActiveParticle, act->NumActiveParticle, 0);
+    /* Build the queue once, since it is really 'all black holes' and similar for all treewalks.
+     * Allocate high so we can keep the tree we build after making the list.*/
+    treewalk_build_queue(tw_dynfric, act->ActiveParticle, act->NumActiveParticle, 0, 1);
```
Comment (rainwoodman) [inline libgadget/blackhole.c:529]: Question 1: Does the force_tree_rebuild_mask alloc-high, such that build_queue must also alloc-high?

Question 2: The alloc_high knob adds a fragile coupling between build_queue and caller. 
This seems to be the only user of the knob. What if  we do an alloc-high here (on the caller side), and memcpy the queue?

File: libgadget/run.c (hunk 3 above)
Comment (sbird) [reply libgadget/run.c:397]: Yes, this is indeed confusing. The first is true and the second false when tree gravity is off. Then we still need a tree to do the PM. The first being false but the second being true should be impossible (I think): crashing is a legit thing to do there. I will add a comment.

File: libgadget/winds.c [hunk 5]
```diff
@@ -252,8 +255,17 @@ int cmp_by_part_id(const void * a, const void * b)
 
 /* Find the 1D DM velocity dispersion of the winds by running a density loop.*/
 static void
-winds_find_weights(TreeWalk * tw, struct WindPriv * priv, int * NewStars, int NumNewStars, const double Time, const double hubble, ForceTree * tree)
+winds_find_weights(TreeWalk * tw, struct WindPriv * priv, int * NewStars, int NumNewStars, const double Time, const double hubble, ForceTree * tree, DomainDecomp * ddecomp)
 {
+    /* Flags that we need to free the tree to preserve memory order*/
+    priv->tree_alloc_in_wind = 0;
```
Comment (rainwoodman) [inline libgadget/winds.c:261]: The logic here reads like:

1. If there is a tree, use it
2. If there is no tree, build one, use it, then free it.

The usual way of implementing such logic is checking tree == NULL; So checking tree->tree_allocated_flag looks a bit surprising. I guess it is complicated to change the calling convention everywhere.

File: libgadget/blackhole.c (hunk 4 above)
Comment (sbird) [reply libgadget/blackhole.c:529]: 1) The force_tree_rebuild_mask is alloc-low but I want to let the built force tree persist past the end of blackhole so I can reuse it for the winds. So I then can't free build_queue in time.
2) It isn't a bad idea. I was just trying to avoid a memcpy, but this is going to be a small one.

File: libgadget/winds.c (hunk 5 above)
Comment (sbird) [reply libgadget/winds.c:261]: Indeed - it is not the tree that should be == NULL, but the tree->Nodes pointer and that has its own oddities with being offset from Nodes_base.

File: libgadget/blackhole.c (hunk 4 above)
Comment (rainwoodman) [reply libgadget/blackhole.c:529]: I usually weight a small memcpy cheaper than a new function argument. :P

**Comment by sbird** [comment]
A small 256^3 15 Mpc box on Frontera is 30 % faster. Let's see if it does anything to P(k), GSMF, HMF or lyman alpha...

File: libgadget/run.c (hunk 3 above)
Comment (rainwoodman) [reply libgadget/run.c:397]: "still need a tree to do the PM": it sounds like even after the GasTree split, the remaining gravTree can still be refactored into two -- the spatial indexing for PM, and the "moments" for short-range grav; in the future.

File: libgadget/blackhole.c (hunk 4 above)
Comment (sbird) [reply libgadget/blackhole.c:529]: It is fair :)

File: libgadget/run.c (hunk 3 above)
Comment (sbird) [reply libgadget/run.c:397]: Yes such a split would be a good idea. We could save some memory for the gasTree, and reduce peak memory usage during density(). Just needs some messing about because C does not do generics :) It is not worth optimising the case where gravTree is disabled and density is on to avoid an extra rebuild because we almost never run in that mode.

**Comment by sbird** [comment]
Thanks for the review!

**Comment by sbird** [comment]
Looks like there is something wrong: I think the gravitational kick for the hydro-active but gravity-inactive SPH particles is not being computed correctly

[powerspectrum-0.3333.pdf](https://github.com/MP-Gadget/MP-Gadget/files/7825409/powerspectrum-0.3333.pdf)
.

**Comment by sbird** [comment]
Got it: the kick was being applied twice if the particle timestep dropped when the particle was inactive (can happen because atime changes).

**Comment by sbird** [comment]
Still the same: I think I need the Gadget-4 timestepping to make this work.

**Comment by sbird** [comment]
Got it: bin_hydro and bin_gravity were swapped in one place

.

**Comment by sbird** [comment]
(note above was run with this patch so that the double timestepping was enabled): 
diff --git a/libgadget/timestep.c b/libgadget/timestep.c
index 27dd6ac5..cfd19850 100644
--- a/libgadget/timestep.c
+++ b/libgadget/timestep.c
@@ -289,7 +289,7 @@ find_timesteps(const ActiveParticles * act, DriftKickTimes * times, const double
             bin_hydro = bin_gravity;
         /* Enforce that the gravity bin is the hydro bin. This is to ensure the code continues to
          * work while we implement the Hamiltonian timesplitting */
-        bin_gravity = bin_hydro;
+        //bin_gravity = bin_hydro;
         /* Only update if both the old and new timebins are currently active.
          * We know that the shorter hydro timestep is active, but we need to check
          * the gravity timestep.*/

**Comment by sbird** [comment]
Current issues: star formation and cooling code is much slower (by a factor of 2).

**Comment by sbird** [comment]
[bhmf-0.3300.pdf](https://github.com/MP-Gadget/MP-Gadget/files/8682930/bhmf-0.3300.pdf)
[gsmf-0.3300.pdf](https://github.com/MP-Gadget/MP-Gadget/files/8682931/gsmf-0.3300.pdf)
[hmf-0.3300.pdf](https://github.com/MP-Gadget/MP-Gadget/files/8682932/h
[powerspectrum-0.3300.pdf](https://github.com/MP-Gadget/MP-Gadget/files/8682933/powerspectrum-0.3300.pdf)
mf-0.3300.pdf)

**Comment by sbird** [comment]
powerspectrum-0 3300-1
hmf-0 3300-1
gsmf-0 3300-1
bhmf-0 3300-1

**Comment by sbird** [comment]
Mass functions all look good: possibly the 'master branch' had a bug during the test, I will re-run it.

**Comment by sbird** [comment]
Ok - the speed regression in the cooling code is for a fairly straightforward reason: some particles are on shorter timebins now. There is a 'push down' logic which reduces the size of the top timebin so that the hierarchy is less deep. When this happens particles have their cooling evaluated more often, on the shorter timestep (as the hydro timestep is enforced to be shorter than the gravity timestep).
This is done because this code is slower if the timebins do not shrink fast enough (as each bin in the timestep hierarchy is evaluated each timestep it is active). So if the largest bin is bin 20 we can save gravitational evaluations by shrinking the timebin if N(bin 19) > N(bin 20)/3. But the gadget-4 calculation does not account for the fact that we also need to evaluate the hydro and other things twice as often for every hydro particle that shrinks the timebin. So we should be less aggressive at pushing down the bins.

**Comment by sbird** [comment]
Rebased on top of current master

**Comment by sbird** [comment]
Current code completes in 48:54 according to slurm and cpu.txt is:
Step 37818, Time: 0.33, MPIs: 32 Threads: 28 Elapsed: 2900.23
                                2909.65 100.0%        9.42 100.0%        9.42       9.42
 BH                               14.45  0.5%        0.01  0.1%        0.01       0.01
 Cooling                         399.09 13.7%        0.29  3.0%        0.29       0.29
  Build@winds.c:0266               72.58  2.5%        0.00  0.0%        0.00       0.00
  Cooling@sfr_eff.c:0272          179.47  6.2%        0.25  2.7%        0.18       0.29
  StarFormation@sfr_eff.c:0369      134.93  4.6%        0.04  0.4%        0.00       0.11
  Wind@winds.c:0414                12.10  0.4%        0.00  0.0%        0.00       0.00
 Domain                          182.06  6.3%        0.25  2.6%        0.23       0.26
 FOF                              95.91  3.3%        3.64 38.7%        3.64       3.65
PMgrav                          454.69 15.6%        1.24 13.2%        1.22       1.26
 SPH                             594.50 20.4%        0.39  4.1%        0.39       0.39
  Build@run.c:0444                124.82  4.3%        0.00  0.0%        0.00       0.00
  Density                         211.58  7.3%        0.18  1.9%        0.18       0.18
  HmaxUpdate@forcetree.c:1358        4.80  0.2%        0.00  0.0%        0.00       0.00
  Hydro                           253.30  8.7%        0.20  2.2%        0.20       0.20
  Write@checkpoint.c:0037          60.39  2.1%        1.54 16.4%        1.54       1.54
 Timeline                         97.48  3.4%        0.01  0.1%        0.01       0.03
 Tree                            942.01 32.4%        1.99 21.2%        1.99       1.99
  Build                            34.70  1.2%        0.04  0.4%        0.04       0.04
   Moments@forcetree.c:0127          6.89  0.2%        0.01  0.1%        0.01       0.01
   Nodes@forcetree.c:0220           27.81  1.0%        0.03  0.3%        0.03       0.03
  Recv@gravshort-tree.c:0163       39.71  1.4%        0.21  2.2%        0.01       0.26
  Wait1@gravshort-tree.c:0164      473.01 16.3%        0.93  9.9%        0.04       1.37
  Walk1@gravshort-tree.c:0159      345.28 11.9%        0.71  7.5%        0.27       1.60

For comparison the master branch is: 51:27
Step 37889, Time: 0.33, MPIs: 32 Threads: 28 Elapsed: 3050.77
 BH                                2.90  0.1%        0.00  0.0%        0.00       0.00
 Cooling                         334.09 10.9%        0.29  3.4%        0.29       0.29
  Cooling@sfr_eff.c:0270          175.52  5.7%        0.25  3.0%        0.18       0.29
  StarFormation@sfr_eff.c:0367      137.62  4.5%        0.04  0.4%        0.00       0.11
  Wind@winds.c:0398                20.95  0.7%        0.00  0.0%        0.00       0.00
 Domain                          173.35  5.7%        0.23  2.8%        0.22       0.25
 FOF                             109.92  3.6%        3.87 46.3%        3.86       3.87
 PMgrav                          458.19 15.0%        1.24 14.9%        1.23       1.27
 SPH                             566.07 18.5%        0.46  5.5%        0.46       0.46
  Build@forcetree.c:0173            0.68  0.0%        0.00  0.0%        0.00       0.00
  Density                         240.96  7.9%        0.20  2.4%        0.20       0.20
  Hydro                           324.42 10.6%        0.26  3.1%        0.26       0.26
  Timeline                         92.96  3.0%        0.02  0.2%        0.01       0.03
Tree                           1177.07 38.5%        0.89 10.6%        0.88       0.89
  Build                           342.64 11.2%        0.02  0.2%        0.01       0.02
   Moments@forcetree.c:0153         75.43  2.5%        0.00  0.0%        0.00       0.00
   Nodes@forcetree.c:0239          267.21  8.7%        0.02  0.2%        0.01       0.02
  HmaxUpdate@forcetree.c:1373        5.22  0.2%        0.00  0.0%        0.00       0.00
  Recv@gravshort-tree.c:0159      100.86  3.3%        0.12  1.5%        0.00       0.17
  Wait1@gravshort-tree.c:0160      360.90 11.8%        0.24  2.9%        0.00       0.40
  Walk1@gravshort-tree.c:0155      306.75 10.0%        0.42  5.0%        0.25       0.66

Notice that the tree build time is *much* less, although somewhat compensated by the need to build a tree in cooling for the winds. The tree walk imbalance time is a little worse: we could further speed this up at the cost of some memory by re-using the accelerations from the first half-step. The hydro and density modules are faster because they only work on a tree containing gas particles.

**Comment by sbird** [comment]
At this time the code has a few potential optimisations, but no bugs. Let's test it a bit more and then merge it if it passes.

**Comment by sbird** [comment]
Requesting reviews: this is the hierarchical gravity timestepping from Gadget-4.

**Comment by sbird** [comment]
Build is failing because conda has apparently decided to take over the C compiler variables

**Comment by sbird** [comment]
Compiler is broken in github - not sure why.

File: libgadget/gravshort.h [hunk 6]
```diff
@@ -43,13 +43,19 @@ struct GravShortPriv {
...
     /* Matter density in internal units.
      * rho_0 = Omega0 * rho_crit
      * rho_crit = 3 H^2 /(8 pi G).
      * This is (rho_0)^(1/3) ,
      * Note: should account for
      * massive neutrinos, but doesn't. */
     double cbrtrho0;
+    /* Whether to calculate the gravitational potential.
+     * Note this is only useful on steps where all particles are active*/
+    int CalcPotential;
```
Comment (rainwoodman) [inline libgadget/gravshort.h:56]: The comment is a bit confusing -- you meant CalcPotential is True only on steps where all particles are active?

If so, perhaps rename this to AllParticlesAreActive?

File: libgadget/slotsmanager.h [hunk 7]
```diff
@@ -46,6 +46,13 @@ struct bh_particle_data {
     /*******************************************************/
     double DragAccel[3];
     double DFAccel[3];
+    MyFloat FullGravAccel[3]; /* Short-range tree acceleration at the most recent timestep
```
Comment (rainwoodman) [inline libgadget/slotsmanager.h:49]: Does this include the PM / long range accel?

File: libgadget/tests/test_forcetree.c [hunk 8]
```diff
@@ -226,7 +226,9 @@ static void do_tree_test(const int numpart, ForceTree tb, DomainDecomp * ddecomp
     /*Time creating the nodes*/
     double start, end;
     start = MPI_Wtime();
-    int nodes = force_tree_create_nodes(tb, numpart, ALLMASK, ddecomp, 0);
+    ActiveParticles Act = {0};
+    Act.NumActiveParticle = numpart;
```
Comment (rainwoodman) [inline libgadget/tests/test_forcetree.c:230]: Is it necesary to set numactiveparticle on Act = {0}? I noticed above the NumActiveParticle line is omitted.

**Comment by rainwoodman** [comment]
I'll take a look at the CI build error. Looks like I need to upgrade my laptop's OS first in order to reproduce the github workflow locally. (with act). Hopefully it is related to how conda changed there cross-platform compiler configurations on Linux.

File: gadget/params.c [hunk 9]
```diff
@@ -106,6 +106,8 @@ create_gadget_parameter_set()
     param_declare_double(ps, "BHOpeningAngle", OPTIONAL, 0.175, "Barnes-Hut opening angle. Alternative purely geometric tree opening angle. Lower values are more accurate.");
     param_declare_double(ps, "TreeRcut", OPTIONAL, 6, "Number of mesh cells at which we cease walking.");
     param_declare_int(ps, "TreeUseBH", OPTIONAL, 2, "If 1, use Barnes-Hut opening angle rather than the standard Gadget acceleration based opening angle. If 2, use BH criterion for the first timestep only, before we have relative accelerations.");
+    param_declare_int(ps, "HiSplitGravityOn", OPTIONAL, 0, "This flag enables the momentum conserving hierarchical timestepping, where only active particles gravitate, from Gadget 4, for the short-range gravity, and splits the hydro and gravitational timesteps.");
```
[the remark this answers is not in the document]
Comment (sbird) [reply gadget/params.c:109]: I started with "HierarchicalGravityOn" but I kept mis-spelling "Hierarchical" and I suspected so would everyone else. Maybe HierarchGravityOn? Or SplitTimesteppingOn?

File: libgadget/forcetree.c [hunk 10]
```diff
@@ -106,74 +106,52 @@ static void force_validate_nextlist(const ForceTree * tree)
...
 
     if(force_tree_allocated(tree)) {
         force_tree_free(tree);
     }
     walltime_measure("/Misc");
 
-    *tree = force_tree_build(PartManager->NumPart, ALLMASK, ddecomp, HybridNuGrav, DoMoments, EmergencyOutputDir);
+    *tree = force_tree_build(ALLMASK, ddecomp, act, HybridNuGrav, DoMoments, EmergencyOutputDir);
 
-    event_listen(&EventSlotsFork, force_tree_eh_slots_fork, tree);
```
[the remark this answers is not in the document]
Comment (sbird) [reply libgadget/forcetree.c:152]: It was to add a newly created star particle to the tree. But now we rebuild the tree more often and in smaller chunks so it is no longer needed. Actually it was not needed before, because the only tree walk we did after star formation was the winds, which do not use star particles. Also it didn't really work reliably: it could only add particles to the lowest level of the tree.

File: libgadget/tests/test_forcetree.c (hunk 8 above)
Comment (sbird) [reply libgadget/tests/test_forcetree.c]: That is a good point - it should not be necessary at the moment but I have been trying to do it in case that changes. This should be a proper function so it is consistent everywhere, and now it is.

**Comment by sbird** [comment]
Force-push to pick up the fix to CI.

File: libgadget/slotsmanager.h (hunk 7 above)
Comment (sbird) [reply libgadget/slotsmanager.h:49]: It does not...I will rename to make that clearer.

File: libgadget/gravshort.h (hunk 6 above)
Comment (sbird) [reply libgadget/gravshort.h:56]: I reworded the comment: it is not quite AllParticlesAreActive because we still need the potential when hierarchical gravity is off.

File: libgadget/blackhole.c [hunk 11]
```diff
@@ -600,10 +600,31 @@ blackhole(const ActiveParticles * act, double atime, Cosmology * CP, ForceTree *
         return;
     }
 
+    /* Move the working set high so we can keep the tree we build after making the list and still free this active set.*/
+    ActiveBlackHoles = mymalloc2("activeBH", NumActiveBlackHoles * sizeof(int));
```
[the remark this answers is not in the document]
Comment (sbird) [reply libgadget/blackhole.c:597]: It isn't exactly - it is used as the argument to all the later treewalks. I keep it in a separate array because it is weird to be passing a subarray of tw_dynfric around.

**Comment by sbird** [comment]
Some changes were found at z=1 in the wind model when hierachical gravity was switched on and off: this might be fixed by using the correct predicted velocities for the DM velocity dispersion. Unfortunately bloats the size of struct particle_data at the moment.

**Comment by sbird** [comment]
Current bug: 
[ 000975.26 ] Begin Step 512, Time: 0.0664731 (24c000), Redshift: 14.0437, Nf = 00000000314182, Systemstep: 0.000107811, Dloga: 0.00162187, status:
[ 000975.26 ] TotNumPart: 0000033554433 SPH 0000016777216 BH 0000000000 STAR 0000000000001
[ 000975.26 ] Occupied:      0      1      2      3      4      5 dt
[ 000975.26 ]  bin=15   16467518   16772733      0      0      0      0 0.00648748
[ 000975.26 ] X bin=14    309674     4454      0      0      1      0 0.00324374
[ 000975.26 ] X bin=13      24      29      0      0      0      0 0.00162187
[ 000975.26 ]        -----------------------------------
[ 000975.26 ] Total:     309698     4483      0      0      1      0 Sum:    314182
[ 000975.26 ] Tree construction for types: 1.
[ 000975.27 ] Tree constructed (type mask: 1) with 33576573 particles. First node 3670016, number of nodes 533848, first pseudo 5569500. NTopLeaves 1156
[ 000975.27 ] Treewalk DENSITY iter 0: total part 309698 max/MPI: 29821 min/MPI: 10521 balance: 1.54065.
[ 000975.30 ] Max ngb=117.35, min ngb=107.335
[ 000975.30 ] Treewalk DENSITY iter 1: total part 5256 max/MPI: 623 min/MPI: 146 balance: 1.8965.
[ 000975.30 ] Treewalk HYDRO iter 0: total part 309698 max/MPI: 29821 min/MPI: 10521 balance: 1.54065.
[ 000975.34 ] Tree constructed with 314180 particles (moments: 1). First node 3670016, number of nodes 15448, first pseudo 5569500. NTopLeaves 1156
[ 000975.34 ] Treewalk GRAVTREE iter 0: total part 314180 max/MPI: 30295 min/MPI: 10676 balance: 1.54281.
[ 000975.34 ] Tree constructed with 53 particles (moments: 1). First node 3670016, number of nodes 5656, first pseudo 5569500. NTopLeaves 1156
[ 000975.34 ] Treewalk GRAVTREE iter 0: total part 53 max/MPI: 32 min/MPI: 0 balance: 9.6602.
[ 000975.34 ] ERROR: Task 11: Tried to add 21562944 active particles, more than 21562864 allowed
[11:43](https://tacc-frontera.slack.com/archives/DR2AUBQBZ/p1655232236221199)
see /scratch3/06431/yueyingn/frontera25/L25N256wdpv/slurm-4490573.out

**Comment by sbird** [comment]
Fixed. 
> Current bug: [ 000975.26 ] Begin Step 512, Time: 0.0664731 (24c000), Redshift: 14.0437, Nf = 00000000314182, Systemstep: 0.000107811, Dloga: 0.00162187, status: [ 000975.26 ] TotNumPart: 0000033554433 SPH 0000016777216 BH 0000000000 STAR 0000000000001 [ 000975.26 ] Occupied: 0 1 2 3 4 5 dt [ 000975.26 ] bin=15 16467518 16772733 0 0 0 0 0.00648748 [ 000975.26 ] X bin=14 309674 4454 0 0 1 0 0.00324374

**Comment by sbird** [comment]
It runs astrid! 30% faster! It gives the same results on a small test and a large run.

gsmf-0 6000-1

bhmf-0 6000-1
hier-gsmf-0 2500-1

Merging!
